src/mongoc/mongoc-topology-description.c

```diff
@@ -0,0 +1,996 @@
+ *           from node's, node cannot be the primary.
+ *       Now that we know this is the primary:
+ *          -If any hosts, passives, or arbiters in node's description aren't
+ *           in the cluster, add them as UNKNOWN servers and begin monitoring.
+ *          -If the cluster has any servers that aren't in node's description
+ *           remove them and stop monitoring.
+ *       Finally, check the cluster for the new primary.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       Changes to the cluster, possible removal of cluster nodes.
+ *
+ *--------------------------------------------------------------------------
+ */
+static void
+_mongoc_topology_description_update_rs_from_primary (mongoc_topology_description_t *topology,
+                                                     mongoc_server_description_t   *server)
+{
+   mongoc_server_description_t *server_iter;
+   mongoc_set_t *set = topology->servers;
+   bson_iter_t member_iter;
+   const bson_t *rs_members[3];
+   const char *member_name;
+   int i;
+
+   if (!_mongoc_topology_description_has_server(topology, server->connection_address)) return;
+
+   /*
+    * 'Server' can only be the primary if it has the right rs name.
+    */
+
+   if (!topology->set_name) {
+      int len = strlen(server->set_name) + 1;
+      topology->set_name = bson_malloc (len);
+      memcpy (topology->set_name, server->set_name, len);
+   }
+   else if (topology->set_name != server->set_name) {
+      _mongoc_topology_description_remove_server(topology, server);
+      _mongoc_topology_description_check_if_has_primary(topology, server);
+      return;
+   }
+
+   /*
+    *'Server' is the primary! Invalidate other primaries if found.
+    */
+
+   for (i = 0; i < set->items_len; i++) {
+      server_iter = set->items[i].item;
+      if (server_iter->id != server->id &&
+          server_iter->type == MONGOC_SERVER_RS_PRIMARY) {
+         mongoc_server_description_set_state(server, MONGOC_SERVER_UNKNOWN);
+         // TODO SDAM set other states to defaults?
+      }
+   }
+
+   /*
+    * Begin monitoring any new servers primary knows about.
+    */
+
+   rs_members[0] = &server->hosts;
+   rs_members[1] = &server->arbiters;
+   rs_members[2] = &server->passives;
+
+   for (i = 0; i < 3; i++) {
+      bson_iter_init (&member_iter, rs_members[i]);
+
+      while (bson_iter_next (&member_iter)) {
+         member_name = bson_iter_utf8 (&member_iter, NULL);
+
+         if (!_mongoc_topology_description_has_server(topology, member_name)) {
+            mongoc_topology_description_add_server(topology, member_name);
+         }
+      }
+   }
+
+   /*
+    * Stop monitoring any servers primary doesn't know about.
+    */
+
+   for (i = 0; i < set->items_len; i++) {
+      server_iter = set->items[i].item;
+      if (!mongoc_server_description_has_rs_member(server, server_iter->connection_address)) {
+         _mongoc_topology_description_remove_server(topology, server_iter);
+      }
+   }
```

Comment:
I have some concurrency concerns here that may or may not be valid. Can anything read the topology description in an inconsistent state while we update it here?